planner/core/logical_plan_builder.go

```diff
@@ -2038,6 +2116,180 @@ func (b *PlanBuilder) resolveWindowFunction(sel *ast.SelectStmt, p LogicalPlan)
 	return extractor.aggMapper, nil
 }
 
+// correlatedAggregateResolver visits Expr tree.
+// It finds and collects all correlated aggregates which should be evaluated in the outer query.
+type correlatedAggregateResolver struct {
+	ctx       context.Context
+	err       error
+	b         *PlanBuilder
+	outerPlan LogicalPlan
+
+	// correlatedAggFuncs stores aggregate functions which belong to outer query
+	correlatedAggFuncs []*ast.AggregateFuncExpr
+}
+
+// Enter implements Visitor interface.
+func (r *correlatedAggregateResolver) Enter(n ast.Node) (ast.Node, bool) {
+	switch v := n.(type) {
+	case *ast.SelectStmt:
+		if r.outerPlan != nil {
+			outerSchema := r.outerPlan.Schema().Clone()
+			r.b.outerSchemas = append(r.b.outerSchemas, outerSchema)
+			r.b.outerNames = append(r.b.outerNames, r.outerPlan.OutputNames())
+		}
+		r.err = r.resolveSelect(v)
+		return n, true
+	}
+	return n, false
+}
+
+func (r *correlatedAggregateResolver) resolveSelect(sel *ast.SelectStmt) (err error) {
+	useCache, err := r.collectFromTableRefs(r.ctx, sel.From)
+	if err != nil {
+		return err
+	}
+	// we cannot use cache if there are correlated aggregates inside FROM clause,
+	// since the plan we are building now is not correct and need to be rebuild later.
+	p, err := r.b.buildTableRefs(r.ctx, sel.From, useCache)
```

Comment:
We actually can only visit the AST to know the correlated ones. But build a plan here for simplifying the logic of the codes?

Reply:
I thought it would be quite complex to extract correlated columns from AST without building plan, especially when there's subquery in the FROM clause. Do you have any suggestion on doing that?